src/check/constrain/generate/definition.rs

```diff
@@ -164,85 +165,85 @@ pub fn id_from_var(
-            for tup_exp in &tup_exps {
-                constr.add("type and tuple", &ty_exp, tup_exp, &env);
-            }
+            constr.add("variable with type and...", &ty_exp, &Expected::from(var), &env);
+            constr.add("variable with expression", &ty_exp, &Expected::from(expr), &env);
         }
-        if let Some(expr) = expr {
-            for tup_exp in &tup_exps {
-                constr.add("tuple and expression", &Expected::from(expr), tup_exp, &env);
+        (Some(ty), None) => {
+            for (f_name, (f_mut, name)) in match_name(&identifier, ty, var.pos)? {
+                constr.insert_var(&f_name);
+                let ty = Expected::new(var.pos, &Type { name: name.clone() });
+                env = env.insert_var(mutable && f_mut, &f_name, &ty, &constr.var_mapping);
             }
-        }
-    }
 
-    if let Some(ty) = ty {
-        let ty_exp = Expected::new(var.pos, &Type { name: ty.clone() });
-        constr.add("variable and type", &ty_exp, &Expected::from(var), &env);
-    }
-    if let Some(expr) = expr {
-        let msg = format!("variable and expression: `{}`", expr.node);
-        constr.add(&msg, &Expected::from(var), &Expected::from(expr), &env);
-    }
+            let ty_exp = Expected::new(var.pos, &Type { name: ty.clone() });
+            constr.add("variable with only type", &ty_exp, &Expected::from(var), &env);
+        }
+        (None, Some(expr)) => {
+            let mut temp_names = vec![];
+            let fields = identifier.fields(var.pos)?;
+            for (f_mut, name) in &fields {
+                let temp_name = constr.temp_name();
+                temp_names.push(temp_name.clone());
+
+                constr.insert_var(&name);
+                let ty = Expected::new(var.pos, &Type { name: temp_name.clone() });
+                env = env.insert_var(mutable && *f_mut, &name, &ty, &constr.var_mapping);
+
+                let var = AST::new(var.pos, Id { lit: name.clone() });
+                constr.add("variable with only expression", &Expected::from(&var), &ty, &env);
+            }
 
-    Ok(env)
-}
+            let exp_expr = if temp_names.len() > 1 {
+                // if tuple literal, deconstruct elements in generate stage
+                if let Node::Tuple { elements } = &expr.node {
+                    if elements.len() == temp_names.len() {
+                        for (i, (expr, ty)) in enumerate(elements.iter().zip(&temp_names)) {
+                            let expr_exp = Expected::from(expr);
+                            let expr_ty = Expected::new(expr.pos, &Type { name: ty.clone() });
+
+                            let msg = format!("tuple literal element {}", i);
+                            constr.add(&msg, &expr_ty, &expr_exp, &env);
+                        }
+                    } else {
+                        let msg = format!("Cannot join tuple literal. Expected {} elements, was {}",
```

Comment:
`join` is not descriptive to an end user.
Instead, just say `Expected tuple of {} elements [...]`